Output name for html-files ignored

### Describe the bug

HTML-files are ignored, when defining inputs and their output names.

I define the names:
![image](https://github.com/vitejs/vite/assets/1536598/5afed753-fcf0-4181-8907-ccdc8f4c287d)

I expect the HTML-file to be `nameForHtmlFile.html` and placed directly in `dist`, but it will remain the folder structure from `src` and keep it's original name.
![image](https://github.com/vitejs/vite/assets/1536598/341cd51c-3d72-4927-aa5f-0e95fca6b63e)


### Reproduction

Not available

### Steps to reproduce

1. npm create vite@latest test
2. Choose TypeScript
3. cd test
4. npm install
5. move index.html to a deeply nested folder and add a TypeScript file too.
![image](https://github.com/vitejs/vite/assets/1536598/038d016f-0fa8-449d-ac0d-9757377e26f7)
6. Change vite.config.ts to this:
```ts
import {defineConfig} from 'vite'
import react from '@vitejs/plugin-react'

// https://vitejs.dev/config/
export default defineConfig({
    plugins: [react()],
    build: {
        rollupOptions: {
            input: {
                nameForHtmlFile: "./src/sub/sub/sub/sub/index.html",
                nameForTsFile: "./src/sub/sub/sub/sub/something.ts",
            }
        }
    }
})
```
7. run `npm run build`
8. The index.html-file should be named `nameForHtmlFile.html` and be located in the root of the dist-folder.

### System Info

```shell
System:
    OS: Windows 10 10.0.19045
    CPU: (8) x64 Intel(R) Core(TM) i7-9700 CPU @ 3.00GHz
    Memory: 3.02 GB / 15.78 GB
  Binaries:
    Node: 14.15.0 - C:\Program Files\nodejs\node.EXE
    npm: 6.14.8 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (113.0.1774.57)
    Internet Explorer: 11.0.19041.1566
  npmPackages:
    @vitejs/plugin-react: ^4.0.0 => 4.0.0
    vite: ^4.3.9 => 4.3.9
```


### Used Package Manager

npm

### Logs

> tsc && vite build --debug

  vite:config bundled config file loaded in 280.70ms +0ms
  vite:config using resolved config: {
  vite:config   plugins: [
  vite:config     'vite:build-metadata',
  vite:config     'vite:watch-package-data',
  vite:config     'vite:pre-alias',
  vite:config     'alias',
  vite:config     'vite:react-babel',
  vite:config     'vite:react-refresh',
  vite:config     'vite:modulepreload-polyfill',
  vite:config     'vite:resolve',
  vite:config     'vite:html-inline-proxy',
  vite:config     'vite:css',
  vite:config     'vite:esbuild',
  vite:config     'vite:json',
  vite:config     'vite:wasm-helper',
  vite:config     'vite:worker',
  vite:config     'vite:asset',
  vite:config     'vite:wasm-fallback',
  vite:config     'vite:define',
  vite:config     'vite:css-post',
  vite:config     'vite:build-html',
  vite:config     'vite:worker-import-meta-url',
  vite:config     'vite:asset-import-meta-url',
  vite:config     'vite:force-systemjs-wrap-complete',
  vite:config     'commonjs',
  vite:config     'vite:data-uri',
  vite:config     'vite:dynamic-import-vars',
  vite:config     'vite:import-glob',
  vite:config     'vite:build-import-analysis',
  vite:config     'vite:esbuild-transpile',
  vite:config     'vite:terser',
  vite:config     'vite:reporter',
  vite:config     'vite:load-fallback'
  vite:config   ],
  vite:config   build: {
  vite:config     target: [ 'es2020', 'edge88', 'firefox78', 'chrome87', 'safari14' ],
  vite:config     cssTarget: [ 'es2020', 'edge88', 'firefox78', 'chrome87', 'safari14' ],
  vite:config     outDir: 'dist',
  vite:config     assetsDir: 'assets',
  vite:config     assetsInlineLimit: 4096,
  vite:config     cssCodeSplit: true,
  vite:config     sourcemap: false,
  vite:config     rollupOptions: { input: [Object], onwarn: [Function: onwarn] },
  vite:config     minify: 'esbuild',
  vite:config     terserOptions: {},
  vite:config     write: true,
  vite:config     emptyOutDir: null,
  vite:config     copyPublicDir: true,
  vite:config     manifest: false,
  vite:config     lib: false,
  vite:config     ssr: false,
  vite:config     ssrManifest: false,
  vite:config     ssrEmitAssets: false,
  vite:config     reportCompressedSize: true,
  vite:config     chunkSizeWarningLimit: 500,
  vite:config     watch: null,
  vite:config     commonjsOptions: { include: [Array], extensions: [Array] },
  vite:config     dynamicImportVarsOptions: { warnOnError: true, exclude: [Array] },
  vite:config     modulePreload: { polyfill: true },
  vite:config     cssMinify: true
  vite:config   },
  vite:config   optimizeDeps: {
  vite:config     disabled: 'build',
  vite:config     force: undefined,
  vite:config     include: [ 'react', 'react/jsx-dev-runtime' ],
  vite:config     esbuildOptions: { preserveSymlinks: false }
  vite:config   },
  vite:config   esbuild: { jsxDev: false, jsx: 'automatic', jsxImportSource: undefined },
  vite:config   resolve: {
  vite:config     mainFields: [ 'module', 'jsnext:main', 'jsnext' ],
  vite:config     browserField: true,
  vite:config     conditions: [],
  vite:config     extensions: [
  vite:config       '.mjs',  '.js',
  vite:config       '.mts',  '.ts',
  vite:config       '.jsx',  '.tsx',
  vite:config       '.json'
  vite:config     ],
  vite:config     dedupe: [ 'react', 'react-dom' ],
  vite:config     preserveSymlinks: false,
  vite:config     alias: [ [Object], [Object] ]
  vite:config   },
  vite:config   configFile: 'C:/code/test/vite.config.ts',
  vite:config   configFileDependencies: [ 'C:/code/test/vite.config.ts' ],
  vite:config   inlineConfig: {
  vite:config     root: undefined,
  vite:config     base: undefined,
  vite:config     mode: undefined,
  vite:config     configFile: undefined,
  vite:config     logLevel: undefined,
  vite:config     clearScreen: undefined,
  vite:config     optimizeDeps: { force: undefined },
  vite:config     build: {}
  vite:config   },
  vite:config   root: 'C:/code/test',
  vite:config   base: '/',
  vite:config   rawBase: '/',
  vite:config   publicDir: 'C:\\code\\test\\public',
  vite:config   cacheDir: 'C:/code/test/node_modules/.vite',
  vite:config   command: 'build',
  vite:config   mode: 'production',
  vite:config   ssr: {
  vite:config     format: 'esm',
  vite:config     target: 'node',
  vite:config     optimizeDeps: { disabled: true, esbuildOptions: [Object] }
  vite:config   },
  vite:config   isWorker: false,
  vite:config   mainConfig: null,
  vite:config   isProduction: true,
  vite:config   server: {
  vite:config     preTransformRequests: true,
  vite:config     sourcemapIgnoreList: [Function: isInNodeModules],
  vite:config     middlewareMode: false,
  vite:config     fs: { strict: true, allow: [Array], deny: [Array] }
  vite:config   },
  vite:config   preview: {
  vite:config     port: undefined,
  vite:config     strictPort: undefined,
  vite:config     host: undefined,
  vite:config     https: undefined,
  vite:config     open: undefined,
  vite:config     proxy: undefined,
  vite:config     cors: undefined,
  vite:config     headers: undefined
  vite:config   },
  vite:config   envDir: 'C:/code/test',
  vite:config   env: { BASE_URL: '/', MODE: 'production', DEV: false, PROD: true },
  vite:config   assetsInclude: [Function: assetsInclude],
  vite:config   logger: {
  vite:config     hasWarned: false,
  vite:config     info: [Function: info],
  vite:config     warn: [Function: warn],
  vite:config     warnOnce: [Function: warnOnce],
  vite:config     error: [Function: error],
  vite:config     clearScreen: [Function: clearScreen],
  vite:config     hasErrorLogged: [Function: hasErrorLogged]
  vite:config   },
  vite:config   packageCache: Map(1) {
  vite:config     'fnpd_C:/code/test' => {
  vite:config       dir: 'C:\\code\\test',
  vite:config       data: [Object],
  vite:config       hasSideEffects: [Function: hasSideEffects],
  vite:config       webResolvedImports: {},
  vite:config       nodeResolvedImports: {},
  vite:config       setResolvedCache: [Function: setResolvedCache],
  vite:config       getResolvedCache: [Function: getResolvedCache]
  vite:config     },
  vite:config     set: [Function (anonymous)]
  vite:config   },
  vite:config   createResolver: [Function: createResolver],
  vite:config   worker: {
  vite:config     format: 'iife',
  vite:config     plugins: [
  vite:config       'vite:build-metadata',
  vite:config       'vite:watch-package-data',
  vite:config       'vite:pre-alias',
  vite:config       'alias',
  vite:config       'vite:modulepreload-polyfill',
  vite:config       'vite:resolve',
  vite:config       'vite:html-inline-proxy',
  vite:config       'vite:css',
  vite:config       'vite:esbuild',
  vite:config       'vite:json',
  vite:config       'vite:wasm-helper',
  vite:config       'vite:worker',
  vite:config       'vite:asset',
  vite:config       'vite:wasm-fallback',
  vite:config       'vite:define',
  vite:config       'vite:css-post',
  vite:config       'vite:build-html',
  vite:config       'vite:worker-import-meta-url',
  vite:config       'vite:asset-import-meta-url',
  vite:config       'vite:force-systemjs-wrap-complete',
  vite:config       'commonjs',
  vite:config       'vite:data-uri',
  vite:config       'vite:worker-post',
  vite:config       'vite:dynamic-import-vars',
  vite:config       'vite:import-glob',
  vite:config       'vite:build-import-analysis',
  vite:config       'vite:esbuild-transpile',
  vite:config       'vite:terser',
  vite:config       'vite:load-fallback'
  vite:config     ],
  vite:config     rollupOptions: {},
  vite:config     getSortedPlugins: [Function: getSortedPlugins],
  vite:config     getSortedPluginHooks: [Function: getSortedPluginHooks]
  vite:config   },
  vite:config   appType: 'spa',
  vite:config   experimental: { importGlobRestoreExtension: false, hmrPartialAccept: false },
  vite:config   getSortedPlugins: [Function: getSortedPlugins],
  vite:config   getSortedPluginHooks: [Function: getSortedPluginHooks]
  vite:config } +12ms
vite v4.3.9 building for production...
(node:34380) UnhandledPromiseRejectionWarning: ReferenceError: performance is not defined
    at initTSConfckParseOptions (file:///C:/code/test/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:14042:29)
    at initTSConfck (file:///C:/code/test/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:14033:28)
    at buildEsbuildPlugin (file:///C:/code/test/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:13897:5)
    at Object.resolveBuildPlugins (file:///C:/code/test/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:46300:45)
    at async resolvePlugins (file:///C:/code/test/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:42258:11)
    at async resolveConfig (file:///C:/code/test/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:64242:24)
    at async build (file:///C:/code/test/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:46318:20)
    at async CAC.<anonymous> (file:///C:/code/test/node_modules/vite/dist/node/cli.js:812:9)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:34380) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:34380) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
✓ 0 modules transformed.
✓ built in 16ms
[vite:esbuild] performance is not defined
file: C:/code/test/src/sub/sub/sub/sub/something.ts
error during build:
ReferenceError: performance is not defined
    at initTSConfckParseOptions (file:///C:/code/test/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:14042:29)
    at initTSConfck (file:///C:/code/test/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:14033:28)
    at buildEsbuildPlugin (file:///C:/code/test/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:13897:5)
    at Object.resolveBuildPlugins (file:///C:/code/test/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:46300:45)
    at async resolvePlugins (file:///C:/code/test/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:42258:11)
    at async resolveConfig (file:///C:/code/test/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:64242:24)
    at async build (file:///C:/code/test/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:46318:20)
    at async CAC.<anonymous> (file:///C:/code/test/node_modules/vite/dist/node/cli.js:812:9)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! test@0.0.0 build: `tsc && vite build --debug`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the test@0.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\T217913\AppData\Roaming\npm-cache\_logs\2023-05-31T08_13_33_672Z-debug.log

PI-WINDOWS+T217913@SDC9AB0410 MINGW64 /C/code/test

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/docs/guide/build.md
+++ b/docs/guide/build.md
@@ -114,6 +114,8 @@ export default defineConfig({
 
 If you specify a different root, remember that `__dirname` will still be the folder of your vite.config.js file when resolving the input paths. Therefore, you will need to add your `root` entry to the arguments for `resolve`.
 
+Note that for HTML files, Vite ignores the name given to the entry in the `rollupOptions.input` object and instead respects the resolved id of the file when generating the HTML asset in the dist folder. This ensures a consistent structure with the way the dev server works.
+
 ## Library Mode
 
 When you are developing a browser-oriented library, you are likely spending most of the time on a test/demo page that imports your actual library. With Vite, you can use your `index.html` for that purpose to get the smooth development experience.

